Modules/posixmodule.c

```diff
@@ -1701,7 +1703,8 @@ win32_xstat_impl(const wchar_t *path, struct _Py_stat_struct *result,
         } else
             CloseHandle(hFile);
     }
-    _Py_attribute_data_to_stat(&info, reparse_tag, result);
+    is_link = _Py_is_reparse_link(path, reparse_tag);
```

Comment:
`_Py_is_reparse_link` needs to be called separately in both the `attributes_from_dir` and by-handle branches. We need this information in order to determine whether to traverse. Non-link reparse points should always be traversed (e.g. mount points, or some kind of remote storage implemented by a filter driver), but links should only be traversed if `traverse` is true. So the test becomes `!is_link || traverse`. 

Additional notes on  `win32_xstat_impl`:

`attributes_from_dir` should also be used if `CreateFile` fails with   `ERROR_INVALID_PARAMETER`, which occurs for read-write devices such as `CON` that are required to be opened with either read or write access. In this case `FindFirstFile` will fake the result for us. Also, if it's a reparse point, the call should fail based on the same test as above, i.e. `!is_link || traverse`. 

`get_target_path` can break handling volumes mounted in NTFS folders. It returns the DOS path, which may be just the NTFS directory again if the volume isn't also mounted as a logical drive letter. Why do we need to get the final path just to reopen it when we already have a valid handle with which we can directly call `GetFileInformationByHandle`? I think `get_target_path` should be removed.

We shouldn't make this expensive `GetFileInformationByHandle` call twice (it's implemented as two system calls, one for the basic volume information and one for *all* file information, much of which is ignored). The first call should be `GetFileInformationByHandleEx`, to get the `FileAttributeTagInfo`. This also gets the reparse tag, so `win32_get_reparse_tag` can be eliminated. If it fails with either `ERROR_INVALID_FUNCTION` or `ERROR_INVALID_PARAMETER`, we can assume reparse points aren't supported by the device. Otherwise, if it's determined that we should traverse, simply close and reopen `hFile` without `FILE_FLAG_OPEN_REPARSE_POINT`.  Then pass this handle to a common `GetFileInformationByHandle` call. It's still expensive, but at least the most expensive call only occurs once.

Finally, a minor quibble. Using `FILE_SHARE_READ` in the 2nd open is pointless.  Accessing file metadata (e.g. attributes, security) isn't subject to shared access. The share mode is only for accessing file data (i.e. read, execute, write, append, or delete access).

Reply:
> `attributes_from_dir` should also be used if `CreateFile` fails with `ERROR_INVALID_PARAMETER`

I'm not able to follow this paragraph. Is this not currently done? I.e. with `hFile == INVALID_HANDLE_VALUE` test?

> I think `get_target_path` should be removed.

This function was [introduced here](https://github.com/python/cpython/commit/d25aef55c8b0025dd2ee7de11b526f34ceed6b66), with a commit message explaining why. I'm not in a position where I can judge that against the concerns you raise.

Reply:
Never mind my previous comment, I think I got it.